Feature: Support memory qos with cgroups v2

borgerli opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

Implements [KEP - Support memory qos with cgroups v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos)

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Introducing Memory QoS support with cgroups v2 (Alpha)
The MemoryQoS feature is now in Alpha. This allows kubelet running with cgroups v2 to set memory QoS at container, pod and QoS level to protect and guarantee better memory quality. This feature can be enabled through feature gate MemoryQoS.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [KEP] : [Support memory qos with cgroups v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos)

#### Note
After discussion with KEP owner @xiaoxubeii, this PR leaves the implementation of  kube-reserved / system-reserved cgroups for future.  The effective min boundary of a cgroup is also limited by memory.min values of all ancestor cgroups, and memory.min of all ancestor cgroups need to be set accordingly. But for kube-reserved / system-reserved cgroups, the ancestor cgroups may not be managed by kubelet.

Commit:
feature: support Memory QoS for cgroups v2

xiaoxubeii commented:
/cc
/cc @bobbypage @mrunalp

xiaoxubeii commented:
/ok-to-test

borgerli commented:
/retest

borgerli commented:
/test pull-kubernetes-e2e-gce-storage-slow

borgerli commented:
/retest

borgerli commented:
The first commit of this PR,"add unified on cri to support cgroup v2", is being reviewed in another PR: https://github.com/kubernetes/kubernetes/pull/102578. 
I'll rebase after PR102578 merged.

borgerli commented:
/test pull-kubernetes-e2e-aks-engine-windows-containerd

borgerli commented:
/test pull-kubernetes-e2e-aks-engine-windows-containerd

xiaoxubeii commented:
lgtm 
ping @bobbypage @mrunalp for further review.

borgerli commented:
/retest

borgerli commented:
/retest

bobbypage commented:
Closely related change to CRI to support unified - https://github.com/kubernetes/kubernetes/pull/102578

xiaoxubeii commented:
> Closely related change to CRI to support unified - #102578
Yes, I think this pr has rebased for that.

bobbypage commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -89,6 +98,40 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerConfig(container *v1.C
+
+		memoryHigh := int64(0)
+		if memoryLimit != 0 {
+			memoryHigh = int64(float64(memoryLimit) * memoryThrottlingFactor)
```
is it desirable to use the the same `memoryThrottlingFactor` for both calculating `memory.high` when memoryLimit exists as well as when memoryLimit does not exist and using node allocatable memory? Perhaps we want to have two different throttling factors? 

/cc @mrunalp

xiaoxubeii replied:
I think one throttling factor makes it easier for understanding. When container sets the `memory_limit`, throttling is from the container perspective. If `memory limit` is not set, then it is from the global node perspective.

mrunalp replied:
I think this is okay for now but can revisit in next iteration.

bobbypage commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -89,6 +98,40 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerConfig(container *v1.C
+		if memoryRequest != 0 {
+			unified[cm.MemoryMin] = strconv.FormatInt(memoryRequest, 10)
+		}
+
```
can we add a comment here to explain logic from KEP:

i.e.

> If container sets limits.memory, we set memory.high=pod.spec.containers[i].resources.limits[memory] * memory throttling factor for container level cgroup if memory.high>memory.min

> If container doesn't set limits.memory, we set memory.high=node allocatable memory * memory throttling factor for container level cgroup

borgerli replied:
fixed

bobbypage commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -434,6 +438,13 @@ func (m *cgroupManagerImpl) toResources(resourceConfig *ResourceConfig) *libcont
 			Limit:    uint64(0),
 		})
 	}
+	// only for cgroup v2
```
one thing that may become confusing is when `Unified` should be used as opposed to existing `Resources` config from runc libcontainer. As I understand, some of the existing properties like `Memory` (https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/opencontainers/runc/libcontainer/configs/cgroup_linux.go#L49) already map to existing cgroupv2 items like `memory.Max`. 

However for the rest of the cgroupv2 options that are not implicitly converted underneath in `Resources` struct, `Unified` should be used instead. Is my understanding here correct?

/cc @giuseppe

mrunalp replied:
Yes, that's right.

giuseppe replied:
ideally unified is used for all the resources when running on cgroup v2.  It doesn't make difference for the memory.max limit, but for e.g. for the cpu controller you can specify the correct setting without relying on the conversions performed by the OCI runtime

bobbypage commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -434,6 +438,13 @@ func (m *cgroupManagerImpl) toResources(resourceConfig *ResourceConfig) *libcont
 		})
 	}
+	// only for cgroup v2
+	if resourceConfig.Unified != nil {
```
does this need to be guarded with `IsCgroupV2` check?

borgerli replied:
fixed

bobbypage commented on pkg/kubelet/cm/helpers_linux_test.go:
```diff
@@ -622,3 +622,211 @@ func TestHugePageLimits(t *testing.T) {
 	}
 }
+
+func TestResourceConfigForPodWithEnforceMemoryQoS(t *testing.T) {
```
does it make sense to add some unit tests for the configuration of the QoS cgroup classes / configuration of the container resource config?

borgerli replied:
fixed.
1. QoS cgroup: add new test file: pkg/kubelet/cm/qos_container_manager_linux_test.go
2. container: pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go::TestGenerateContainerConfigWithMemoryQoSEnforced

bobbypage replied:
Nice, thanks for adding these. Do we also have test for node container manager changes? i.e. as described here https://github.com/kubernetes/kubernetes/pull/102970/files#r661921645

borgerli commented on pkg/kubelet/cm/node_container_manager_linux.go:
```diff
@@ -131,9 +134,20 @@ func (cm *containerManagerImpl) enforceNodeAllocatableCgroups() error {
+			if rp.Unified == nil {
+				rp.Unified = make(map[string]string)
+			}
+			rp.Unified[MemoryMin] = strconv.FormatInt(*rp.Memory, 10)
```
This is to enforce the memory qos for node according to kube-reserved or system-reserved setttings, implementing points 1 & 2 at https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos#node

bobbypage replied:
I see, it would be great to note this as a comment here or perhaps link to the KEP as you did in your comment.


One additional question, is this code path (i.e. `enforceExistingCgroup()`) called specifically only for when `enforce-node-allocatable=kube-reserved` and `enforce-node-allocatable=system-reserved` are set?

borgerli commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -19,19 +19,27 @@ limitations under the License.
 
+const (
+	// The Default memory throttling factor for MemroyQoS
+	memoryThrottlingFactor = 0.8
```
Maybe we could make this configurable in beta based on feedback from community users?

mrunalp replied:
I think making it configurable will help us test with different values more easily but I am fine either ways.

borgerli replied:
Fixed. Now it  can be configured in KubeletConfiguration.

borgerli commented on pkg/kubelet/cm/qos_container_manager_linux.go:
```diff
@@ -290,6 +332,12 @@ func (m *qosContainerManagerImpl) UpdateCgroups() error {
 	}
 
+	// update the qos level cgrougs v2 settings of memory qos if feature enabled
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.MemoryQoS) &&
```
Yes, qosContainerManagerImpl is used only when --cgroups-per-qos is set to true.

borgerli commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:
```diff
@@ -19,19 +19,27 @@ limitations under the License.
 )
 
+const (
+	// The Default memory throttling factor for MemroyQoS
```
fixed

giuseppe commented:
/test pull-kubernetes-node-crio-cgrpv2-e2e

xiaoxubeii commented:
/retest

borgerli commented:
/label api-review

xiaoxubeii commented:
/retest

borgerli commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

liggitt commented:
routing to @smarterclayton for review of kubelet config change per https://github.com/kubernetes/kubernetes/blob/7e00f5d401df08ea4c0c8982594fedade3a1118b/OWNERS_ALIASES#L385-L387

bobbypage commented on pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:
```diff
@@ -144,6 +145,78 @@ func TestGenerateContainerConfig(t *testing.T) {
 	assert.Error(t, err, "RunAsNonRoot should fail for non-numeric username")
 }
 
+func TestGenerateContainerConfigWithMemoryQoSEnforced(t *testing.T) {
```
nit: consider refactoring this to a table driven test. That way it will be more clear for two separate test cases (since you can give them an name) and avoid some of the duplication.

i.e. test case 1 -- validate that pod with requests & limits has memory.min == requests ; memory.high == limits * throttling factor
test case 2 -- validate that pod with requests & no limits has memory.min == requests; memory.high == memoryNodeAllocatable * throttling factor

borgerli replied:
Yes for 
> One additional question, is this code path (i.e. enforceExistingCgroup()) called specifically only for when enforce-node-allocatable=kube-reserved and enforce-node-allocatable=system-reserved are set?

Refactored the test case accordingly, Thanks.

bobbypage replied:
thanks for and info, and thanks for refactoring test, looks much cleaner now.

borgerli commented on pkg/kubelet/apis/config/v1beta1/defaults.go:
```diff
@@ -33,9 +33,10 @@ import (
+	DefaultIPTablesMasqueradeBit  = 14
+	DefaultIPTablesDropBit        = 15
+	DefaultVolumePluginDir        = "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/"
+	DefaultMemoryThrottlingFactor = 0.8
```
fixed.

borgerli commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -409,6 +409,12 @@ type KubeletConfiguration struct {
 	// SeccompDefault enables the use of `RuntimeDefault` as the default seccomp profile for all workloads.
 	SeccompDefault bool
+	// MemoryThrottlingFactor specifies the factor multiplied by the memory limit or node allocatable memory
+	// when setting the cgroupv2 memory.high value to enforce MemoryQoS
```
fixed

borgerli commented:
/retest

borgerli commented:
/retest

borgerli commented:
/retest

borgerli commented:
/retest

borgerli commented:
/retest

borgerli commented:
/retest

borgerli commented:
/retest

borgerli commented:
The check of pull-kubernetes-e2e-gce-alpha-features always failed due to kube-scheduler not running with following errors:
## errors in kube-scheduler  log
I0702 09:43:59.647023       9 flags.go:59] FLAG: --write-config-to=""
profiles[0].pluginConfig[5].args.shape: Required value: at least one point must be specified

## errors in build log
I0702 11:17:01.493] scheduler            Unhealthy   Get "http://127.0.0.1:10251/healthz": dial tcp 127.0.0.1:10251: connect: connection refused   
I0702 11:17:01.493] etcd-1               Healthy     {"health":"true","reason":""}                                                                 
I0702 11:17:01.493] etcd-0               Healthy     {"health":"true","reason":""}                                                                 
I0702 11:17:01.494] controller-manager   Healthy     ok                                                                                            
I0702 11:17:01.494] [0;31mValidation returned one or more failed components. Cluster is probably broken.[0m

borgerli commented:
/retest

borgerli commented:
/test pull-kubernetes-e2e-gce-storage-slow

bobbypage commented on pkg/kubelet/cm/qos_container_manager_linux_test.go:
```diff
@@ -0,0 +1,154 @@
+
+	m.setMemoryQoS(qosConfigs)
+
+	burstableMin := resource.MustParse("384Mi")
```
nit: leave comment of where magic 384Mi comes from (i.e. `128+256` from burstable-pod-1 + burstable-pod-2 memory requests).

bobbypage commented on pkg/kubelet/cm/qos_container_manager_linux_test.go:
```diff
@@ -0,0 +1,154 @@
+	m.setMemoryQoS(qosConfigs)
+
+	burstableMin := resource.MustParse("384Mi")
+	guaranteedMin := resource.MustParse("128Mi")
```
nit: leave comment of where 128Mi comes from (i.e. `guaranteed-pod` memory requests).

bobbypage commented:
Thanks for addressing comments, this looks good to me, thanks for all you efforts.

/lgtm

Would like others to also take a look and provide feedback.

borgerli commented:
> Thanks for addressing comments, this looks good to me, thanks for all you efforts.
> 
> /lgtm
> 
> Would like others to also take a look and provide feedback.

Thank you so much for all the advice.

xiaoxubeii commented:
/lgmt Thanks for contributions. Would like to also take a look? @smarterclayton @thockin @mrunalp

borgerli commented:
@smarterclayton @thockin @mrunalp Could you please help review this PR? Thanks.

xiaoxubeii commented:
/test pull-kubernetes-e2e-gce-alpha-features

xiaoxubeii commented:
/assign @mrunalp

xiaoxubeii commented:
/retest

borgerli commented:
> pull-kubernetes-e2e-gce-alpha-features

According to the job history: https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-gce-alpha-features, this check has been always failed since `Jul 02 10:15:06`.
I opened an issue for this: https://github.com/kubernetes/test-infra/issues/22796

xiaoxubeii commented:
/lgtm

ehashman commented:
/priority important-soon
/triage accepted

borgerli commented:
/test pull-kubernetes-integration

borgerli commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

xiaoxubeii commented:
/lgtm

borgerli commented:
/test pull-kubernetes-node-e2e-containerd

borgerli commented:
/test pull-kubernetes-e2e-kind-ipv6

borgerli commented:
/test pull-kubernetes-node-e2e-containerd

borgerli commented:
/test pull-kubernetes-e2e-aks-engine-windows-containerd

borgerli commented:
/test pull-kubernetes-e2e-gce-alpha-features

xiaoxubeii commented:
/assign @thockin 
/lgtm

thockin commented on pkg/kubelet/apis/config/fuzzer/fuzzer.go:
```diff
@@ -106,6 +107,7 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 				obj.Logging.Format = "text"
 			}
 			obj.EnableSystemLogHandler = true
+			obj.MemoryThrottlingFactor = utilpointer.Float64Ptr(0.8)
```
why a constant and not random?

xiaoxubeii replied:
https://github.com/kubernetes/enhancements/blame/master/keps/sig-node/2570-memory-qos/README.md#L95

> e.g. if a container has `requests.memory=50, limits.memory=100`, and we have a throttling factor of .8, `memory.high` would be 80. if a container had no memory limit specified, we substitute `limits.memory` for `node allocatable memory` and apply the throttling factor of .8 to that value.


According to the kep definition, the default value is 0.8. This is based on practical experience.

thockin replied:
The point of a fuzzer is to try unexpected values.

xiaoxubeii replied:
Thanks for reminding, have updated.

thockin commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -212,5 +212,12 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
 		allErrors = append(allErrors, errs.ToAggregate().Errors()...)
 	}
+
+	if localFeatureGate.Enabled(features.MemoryQoS) {
```
The usual pattern here is:

if gate enabled && nil {
  // error: required value
}
if !nil {
    // check values
}

xiaoxubeii replied:
Updated.

thockin commented on pkg/kubelet/apis/config/validation/validation.go:
```diff
@@ -212,5 +212,12 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error
+
+	if localFeatureGate.Enabled(features.MemoryQoS) {
+		if kc.MemoryThrottlingFactor != nil && (*kc.MemoryThrottlingFactor <= 0 || *kc.MemoryThrottlingFactor > 1.0) {
+			allErrors = append(allErrors, fmt.Errorf("invalid configuration: MemoryThrottlingFactor %v must be greater than 0 and smaller than or equal to 1.0", kc.MemoryThrottlingFactor))
```
use the API-style name `memoryThrottlingFactor` (lower-0case first char)

thockin replied:
s/smaller/less

xiaoxubeii replied:
Updated.

thockin commented on pkg/kubelet/apis/config/fuzzer/fuzzer.go:
```diff
@@ -106,6 +108,7 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 				obj.Logging.Format = "text"
 			}
 			obj.EnableSystemLogHandler = true
+			obj.MemoryThrottlingFactor = utilpointer.Float64Ptr(rand.Float64())
```
LOL>  I made you fix it but when I scroll up, this WHOLE FILE is wrong.  Sorry.

thockin reviewed (commented):
API LGTM

xiaoxubeii commented:
/retest

mrunalp commented:
/approve
for Node. Thanks!

xiaoxubeii commented:
/unassign smarterclayton

ehashman reviewed (commented):
Read through the implementation and SGTM.

Graduation criteria for alpha:
- [x] cgroup_v2 is in Alpha
- [x] Memory QoS is implemented for new feature gate (`MemoryQoS`)
- [ ] Memory QoS is covered by proper tests: We have unit tests but it does not appear there are any e2es?
- [x] Memory QoS supports containerd, cri-o

Given missing e2es I am not sure if this is good to go. Let's at least run this through alpha features and cgroupsv2 coverage.

ehashman commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

borgerli commented:
/retest

borgerli commented:
/test pull-kubernetes-e2e-gce-storage-slow
/test pull-kubernetes-integration

xiaoxubeii commented:
/retest

borgerli commented:
/retest

xiaoxubeii commented:
/test pull-kubernetes-integration

xiaoxubeii commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

xiaoxubeii commented:
> Read through the implementation and SGTM.
> 
> Graduation criteria for alpha:
> 
> * [x]  cgroup_v2 is in Alpha
> * [x]  Memory QoS is implemented for new feature gate (`MemoryQoS`)
> * [ ]  Memory QoS is covered by proper tests: We have unit tests but it does not appear there are any e2es?
> * [x]  Memory QoS supports containerd, cri-o
> 
> Given missing e2es I am not sure if this is good to go. Let's at least run this through alpha features and cgroupsv2 coverage.

Can we use cgv2  e2e in alpha first?

borgerli commented:
/test pull-kubernetes-integration

borgerli commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

xiaoxubeii commented:
/lgtm

borgerli commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

xiaoxubeii commented:
/test pull-kubernetes-node-crio-cgrpv2-e2e

borgerli commented:
/test pull-kubernetes-node-crio-cgrpv2-e2e

borgerli commented:
/test pull-kubernetes-node-crio-cgrpv2-e2e

borgerli commented:
/test pull-kubernetes-node-crio-cgrpv2-e2e

borgerli commented:
The check of `pull-kubernetes-node-kubelet-serial-crio-cgroupv2` always failed due to issue #103579.

borgerli commented:
/test pull-kubernetes-node-crio-cgrpv2-e2e

xiaoxubeii commented:
/assign @derekwaynecarr

thockin commented:
/approve 

based on @bobbypage and @mrunalp and @ehashman reviews.

ehashman commented:
> @ehashman Can we use cgv2 e2e in alpha first? It will take some time to submit e2e PR, which we can make up later.

Given that pull-kubernetes-e2e-gce-alpha-features is passing there is at least _some_ coverage for the feature flag being enabled, but since none of the logic of this PR is triggered without the feature flag enabled *and* cgroupsv2, IMO this needs a test job added. Since none of our available jobs are able to check this right now, I'm hesitant to merge without exercising these code paths in CI. This PR should have had some test-infra changes submitted simultaneously so that we can trigger such a job on a PR. See e.g. https://github.com/kubernetes/test-infra/blob/58ae43ff3e79c2958827883f5f5b0164084bfb23/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml#L377-L415

If we can't get something merged in test-infra by EOD to test this I'll defer to Node leads to make the call since this is alpha, but I really think we need to add a CI job. This hasn't actually run through CI so I don't have confidence that it works.

ehashman commented:
/hold 

to resolve discussion

ehashman commented:
/test pull-kubernetes-node-memoryqos-cgrpv2

ehashman commented:
Job passed! :partying_face: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102970/pull-kubernetes-node-memoryqos-cgrpv2/1413181901090328576/

/hold cancel
/approve

ehashman commented:
/skip

ehashman commented:
/milestone v1.22
